### DISCUSSION TITLE
liturgicalColorNames for 18th October 2021, St Luke, shows white.

### DISCUSSION THREAD

**Discussion by pvanthony** [romcal/romcal#314] [General]
18th October 2021, St Luke, shows white. Should it not be red?
Need advice.

**Comment by tukusejssirs** [comment]
Indeed it should be red. What romcal version do you use exactly? v1.3?

[romcal v3(alpha)](https://romcal.js.org/) does compute the colour correctly.

**Reply by pvanthony** [reply]
Is there some command that I can use in console.log to see the current version number?

**Reply by tukusejssirs** [reply]
In versions prior to `v3`, there is no command that would return the romcal version. In `v3`, you could run `Romcal.getVersion()`.

How did you install romcal? If you installed it using `npm i romcal`, you could open `{path}/node_modules/romcal/package.json` and look at the `version` (probably on the third line).

**Reply by pvanthony** [reply]
opps! I am using version 2. I will compile version 3. 
Thanks for the quick answer.

**Reply by tukusejssirs** [reply]
Note that `v3` is of alpha quality.

**Reply by tukusejssirs** [reply]
And it works totally differently than `v2`, therefore in order to update from `v2` to `v3` you need to update your code. Also the [`docs/`](https://github.com/romcal/romcal/tree/dev/docs) aren’t fully updated yet.

**Reply by pvanthony** [reply]
oh no! So v3 is not backward compatible. Good thing you told me.
I downloaded using git clone and the branch is dev. Is that correct?

After downloading, tried to build dev but failed because of errors.
Did I get the correct branch?

**Reply by tukusejssirs** [reply]
> oh no! So v3 is not backward compatible. Good thing you told me.

We follow [semantic versioning](https://semver.org/), therefore whenever the major version number changes, you can be sure there are some backward-incompatible changes.

>  I downloaded using git clone and the branch is dev. Is that correct?

Yes, `dev` is the development branch (`alpha` channel).

> After downloading, tried to build dev but failed because of errors.

Running `npm install` and `npm run build` should do the job, but the tests are not yet fixed and those tests are run as the pre-build step. Try running `node -r ts-node/register scripts/build.ts` instead of `npm run build` in the root of your local romcal folder (where you have cloned the repo). This way you can skip the tests until they are fixed.

**Reply by emagnier** [reply]
No necessary to do a `git clone`. Doing that is only for advanced contributions (if you want to modify the romcal codebase or data).

Option B below is easier.

### Option A: 

Just do an `npm install romcal@alpha`.

You can follow the steps here: https://github.com/romcal/romcal#getting-started
But do not forget the `@alpha` at the end, it's actually missing from the documentation (we have to add it).

Since `v3`, calendars are separated to the core library, so eventually you also have to install the calendar you need on your project.
See _calendar installation_: https://github.com/romcal/romcal#calendars-installation
All the calendars are listed here: https://github.com/romcal/romcal/blob/dev/docs/calendar-plugins.md

For example, to install the latest version of romcal and the General Roman Calendar:

```bash
# At the root of your project:

# only if you don't have a 'package.json' at the root of your project
npm init -y

# then...
npm install romcal@alpha
npm install @romcal/calendar.general-roman
``` 

Then, if you are embedding the library from a `<script>` in hour HTML page, you will find the corresponding files in:
- node_modules/romcal/dist/iife/romcal.js
- node_modules/@romcal/calendar.general-roman/iife/en.js

Otherwise, other possibilities to embed romcal are described in the doc: https://github.com/romcal/romcal#basic-samples

### Option B

Or more easily, you can use a CDN:
```html
<script src="https://unpkg.com/romcal@alpha/dist/iife/romcal.js"></script>
<script src="https://unpkg.com/@romcal/calendar.general-roman/iife/fr.js"></script>
```

### Then...

The options to consume romcal data are a bit different.

```js
// Initialize the romcal library
const romcal = new Romcal({
  localizedCalendar: GeneralRoman_En,
  scope: 'gregorian',
});

// To get romcal data
romcal.generateCalendar(2030).then((data) => {
  console.log(data);
});
```

The JS `Object` for each of the days have minor differences, so if you don't have advanced usages, you might not see any differences.

**But**, the whole collection of data result are now outputted differently. If you don't want to change your code logic, you can convert the results to be similar than romcal `v2`:

```js
// To get romcal data like romcal v2
// Note that if you don't provide year in generateCalendar(), it will take the current year
romcal.generateCalendar().then((data) => {
  const dataLikeRomcalV2 = Object.values(data).flat();
  const dataLikeRomcalV2_withoutOptionalMemorials = Object.values(data).flatMap((arr) => arr[0]);

  console.log(dataLikeRomcalV2);
  console.log(dataLikeRomcalV2_withoutOptionalMemorials);
});
```

Thank you for your patience!

**Reply by pvanthony** [reply]
I need to change the country data. For example, like singapore.ts. Then I need to compile, correct?

**Reply by tukusejssirs** [reply]
> I need to change the country data. For example, like singapore.ts.

Well, we need to _create_ a calendar for Singapore, as currently it is not in romcal.

You can find all currently supported calendars are [these](https://github.com/romcal/romcal/tree/dev/lib/particular-calendars) or you could use `Romcal.CALENDAR_VAR_NAMES` in romcal `v3` to get a list (an array) of the available calendars.

Anyway, I could help you create `singapore.ts`, but I need the sources (i.e. the calendar from official liturgical books, e.g. Roman Missal in your language, but that might be outdated a bit if a new celebration was inserted into the calendar after the book was published, therefore an Ordo / liturgical calendar guide might be helpful too). Of course, I don’t speak any Asian language, therefore I would need help in this area too. :wink: Nevertheless, it might be helpful to know the continental patrons (patrons of Asia) and diocesian patrons, which should be in their respective calendars (continental celebrations in continental calendars, like `asia.ts`, diocesan in `{country}.{diocese}.ts`), therefore it might be good to know the ecclesiastical structures too.

---

Once you mentioned you also work with `sri-lanka.ts`, which we already have in romcal (the accuracy is questionable; it might require some corrections), for which you don’t need to create that calendar, simply use either of the following:

```js
// From unpkg.com
// General pattern: https://unpkg.com/@romcal/calendar.{calendar}@3.0.0-alpha.0/iife/{locale}.js
https://unpkg.com/@romcal/calendar.sri-lanka@3.0.0-alpha.0/iife/en.js

// From locally built file
// Note: The constant names must be `{Calendar}_{Locale}`, e.g. `Singapore_En`.
const {Slovakia_Sk} = require('dist/bundles/slovakia/cjs/sk')
```

**Reply by emagnier** [reply]
@tukusejssirs, just a small clarification, when you `require(...)` a calendar, no needs to add `/cjs/sk` at the end. You can just do:

```js
const {Slovakia_Sk} = require('dist/bundles/slovakia')
```

The right build format (`cjs` or `esm`) is automatically taken (the different builds are specified in the `package.json` of the calendar). And there is an `index.js` that refers to all available locales for this calendar. The only situation where you need to provide the full path is when you are using an `iife` build, i.e. inside a `<script>` tag in an HTML page.

**Reply by tukusejssirs** [reply]
> no needs to add `/cjs/sk` at the end.

Good to know! :wink:

**Reply by emagnier** [reply]
@pvanthony, indeed, if you plan to add a new `singapore` calendar (`lib/particular-calendars/singapore.ts`), you might have to compile/build the project. But this isn't required. Only recommended if you want to test quickly the calendar or data you are modifying.

From a `git clone` of the romcal project, take care that you are on the latest version of the `dev` branch (if you just made a `git clone` you will be already up-to-date). Then do `npm install` and `npm run build`. Then you will find the built library and calendars in the `dist` directory.

You also need to have the Node.js version `>14.16` or `>16` installed on your computer (to check that, do `node --version`).
Optional (but useful): to make Node.js used versions managed automatically, without having to take care of that, you can install [Volta](https://volta.sh/). Once this tool is installed on your machine, it will automatically set up the right recommended Node.js and NPM version to run romcal.

**Comment by pvanthony** [comment]
I would like to use Option B. It is much simpler. I will change my code to use version 3. That would be better moving forward I think.
If I create the singapore.ts, can I still use Option B and "overlay" the singapore.ts file? Can it work like that or I have to compile romcal?

**Reply by tukusejssirs** [reply]
Well, you need to create it in `lib/particular-calendars` in your local copy of romcal, _then_ you definitely need to build (compile) it using `npm run build`. I can help you with it, but need the calendar proper of Singapore. :wink:

Anyway, remind me please what is the _option B_ exactly.

---

> […] you might have to compile/build the project. But this isn't required.

@emagnier, how is it _not_ required to compile `singapore.ts`? A TS file always needs to be compiled, although @pvanthony is not required to commit the code back to romcal.

As I see it, the easiest solution is to create `lib/particular-calendars/singapore.ts` and re-build romcal.

**Reply by pvanthony** [reply]
> Anyway, remind me please what is the _option B_ exactly.
From this answer, @emagnier mentions "Option B".
https://github.com/romcal/romcal/discussions/314#discussioncomment-1497375

**Reply by tukusejssirs** [reply]
Thanks, @pvanthony.

As for the option B, I am not sure if it is feasable if you need to create out-of-source calendar (`singapore.ts`). Maybe @emagnier can think about a solution for this, but as I see it, it that case you need to need to compile/build it and that is easier done within romcal repo cloned on your computer.

However, _if_ we create `singapore.ts` and push it into romcal, then you can use the packages from unpkg.com, as they would then have that calendar included.

**Reply by emagnier** [reply]
If you are modifying romcal data (calendar or localization), this "option B" will work as soon as [1] your contribution is merged into the `dev` branch (from a GitHub Pull Request), and [2] a new version of romcal is published on NPM (which make it also available on CDNs, like unpkg.com). 

This is the easiest way, as you don't necessary have to build romcal on your computer, since it's made automatically when a new romcal version is released.

---

Otherwise, you will have to build romcal if:
- you made additions into romcal data (calendar or localization), and you want to embed it quickly on your project.
- you want to test it before submitting your contribution to this GitHub repository.

**Reply by tukusejssirs** [reply]
Thanks, @emagnier, for confirming what I said.

Therefore, @pvanthony, could you get a copy of the Singapore calendar at least from the Roman Missal (in your language)? That calendar is usually printed on the first pages, usually after the General Instruction of the Roman Missal. Even simple photos taken by your phone would suffice for now (of course, your knowledge of the language is requured :wink:).

**Reply by pvanthony** [reply]
> > […] you might have to compile/build the project. But this isn't required.
> 
> @emagnier, how is it _not_ required to compile `singapore.ts`? A TS file always needs to be compiled, although @pvanthony is not required to commit the code back to romcal.
> 
> As I see it, the easiest solution is to create `lib/particular-calendars/singapore.ts` and re-build romcal.

Thank you for the definitive answer. I was just being lazy. haha.

With the regards to the singapore.ts, let me double check and make sure it is correct before I post here for advice.

Please note I will be late because I will be modifying the javascript for version 3.

**Reply by tukusejssirs** [reply]
> Thank you for the definitive answer.

Anytimes. :wink:

> With the regards to the singapore.ts, let me double check and make sure it is correct before I post here for advice.

Right. Then I think you need to combine multiple sources, unless you find a current list of celebrations celebrated in Singapore (and its dioceses), i.e. the so-called _proper_. By _multiple sources_ I mean sources like:
- Roman Missal,
- ordos (or guides published annually that contains liturgical information about all celebrations in a particular year, which include Mass readings and other stuff);
- in case of recently added celebrations, you might need to follow some articles published by the liturgical commitee at the episcopal conference of your Singapore.

Actually, the liturgical commitee should be able to provide the current proper if they wanted to. :wink:

> Please note I will be late because I will be modifying the javascript for version 3.

No need to hurry. :wink:
